content/300-guides/050-database/052-extend-client.mdx

```diff
@@ -0,0 +1,445 @@
+
+Let's build a signup function below that requires user and team information to create rows in your database.
+
+To make this type-safe, we'll utilize the exported types in `@prisma/client` along with some built-in types that Typescript provides.
+
+This example also demonstrates how you can use the built-in `Omit` type to exclude fields from the input, further restraining what the signup function will accept.
+
+<TabbedContent tabs={[<FileWithIcon text="TypeScript" icon="code"/>, <FileWithIcon text="JavaScript (Node.js)" icon="code"/>]}>
+
+<tab>
+
+```tsx highlight=5-9;normal
+import { PrismaClient, Prisma, User, Team } from '@prisma/client'
+
+const prisma = new PrismaClient()
+
+type Signup = {
+  // accept all team and user fields except id
+  team: Omit<Prisma.TeamCreateArgs['data'], 'id'>
+  user: Omit<Prisma.UserCreateArgs['data'], 'id'>
+}
+
+async function signup(input: Signup): Promise<User> {
+  const team = await prisma.team.create({
+    data: {
+      ...input.team,
+      users: {
+        create: input.user,
+      },
+    },
+    include: {
+      users: true,
+    },
+  })
+  return team.users[0]
+}
+
+// Run the signup function
+async function main() {
+  const { team, user } = await signup({
+    team: {
+      name: 'Super Coffee',
+    },
+    user: {
+      firstName: 'Alice',
+      lastName: 'Wonderland',
+      email: 'alice@prisma.io',
+      password: 'everything is awesome',
+    },
+  })
+}
+```
+
+</tab>
+
+<tab>
+
+```tsx
+import { PrismaClient } from '@prisma/client'
+
+const prisma = new PrismaClient()
+
+// When you first signup, you create an initial user and a team of one
+// with you as the first user.
+async function signup(input) {
+  const team = await prisma.team.create({
+    data: {
+      ...input.team,
+      users: {
+        create: input.user,
+      },
+    },
+    include: {
+      users: true,
+    },
+  })
+  return team.users[0]
+}
+
+// Run the signup function
+async function main() {
+  const { team, user } = await signup({
+    team: {
+      name: 'Super Coffee',
+    },
+    user: {
+      firstName: 'Alice',
+      lastName: 'Wonderland',
+      email: 'alice@prisma.io',
+      password: 'everything is awesome',
+    },
+  })
+}
+```
+
+</tab>
+
+</TabbedContent>
+
+## Validating User Input
+
+Some data must be validated at runtime. For example, there's no way to type-check that your user's email contains an `@` symbol at build time. To ensure that the data you get is the data you expect, you can use runtime validation.
+
+The Node.js ecosystem provides a number of high-quality, easy-to-use validation libraries you can choose from including: [joi](https://github.com/sideway/joi), [validator.js,](https://github.com/validatorjs/validator.js) [yup](https://github.com/jquense/yup), [zod](https://github.com/colinhacks/zod) and [superstruct](https://github.com/ianstormtaylor/superstruct).
```

Comment:
@matthewmueller 

If you already decided prisma client's embedded validator (joi, yup, zod, superstruct or etc), Could you write that here ?

Reply:
Prisma currently has no opinions about custom validation, so you can choose any of the options above.

Reply:
Got it. Thanks for replying.